apps/typegpu-docs/src/content/docs/guides/tgpu-cli.mdx

```diff
@@ -0,0 +1,521 @@
+  fn main(@builtin(global_invocation_id) GlobalInvocationID : vec3u) {
+    var index = GlobalInvocationID.x;
+
+    var vPos = particlesA.particles[index].pos;
+    var vVel = particlesA.particles[index].vel;
+    var cMass = vec2(0.0);
+    var cVel = vec2(0.0);
+    var colVel = vec2(0.0);
+    var cMassCount = 0u;
+    var cVelCount = 0u;
+    var pos : vec2f;
+    var vel : vec2f;
+
+    for (var i = 0u; i < arrayLength(&particlesA.particles); i++) {
+      if (i == index) {
+        continue;
+      }
+
+      pos = particlesA.particles[i].pos.xy;
+      vel = particlesA.particles[i].vel.xy;
+      if (distance(pos, vPos) < params.rule1Distance) {
+        cMass += pos;
+        cMassCount++;
+      }
+      if (distance(pos, vPos) < params.rule2Distance) {
+        colVel -= pos - vPos;
+      }
+      if (distance(pos, vPos) < params.rule3Distance) {
+        cVel += vel;
+        cVelCount++;
+      }
+    }
+    if (cMassCount > 0) {
+      cMass = (cMass / vec2(f32(cMassCount))) - vPos;
+    }
+    if (cVelCount > 0) {
+      cVel /= f32(cVelCount);
+    }
+    vVel += (cMass * params.rule1Scale) + (colVel * params.rule2Scale) + (cVel * params.rule3Scale);
+
+    // clamp velocity for a more pleasing simulation
+    vVel = normalize(vVel) * clamp(length(vVel), 0.0, 0.1);
+    // kinematic update
+    vPos = vPos + (vVel * params.deltaT);
+    // Wrap around boundary
+    if (vPos.x < -1.0) {
+      vPos.x = 1.0;
+    }
+    if (vPos.x > 1.0) {
+      vPos.x = -1.0;
+    }
+    if (vPos.y < -1.0) {
+      vPos.y = 1.0;
+    }
+    if (vPos.y > 1.0) {
+      vPos.y = -1.0;
+    }
+    // Write back
+    particlesB.particles[index].pos = vPos;
+    particlesB.particles[index].vel = vVel;
+  }
+  ```
+  </details>
+  </TabItem>
+  <TabItem label="sprite.wgsl">
+  <details>
+  <summary>Click to see the content</summary>
+  ```wgsl
+  struct VertexOutput {
+    @builtin(position) position : vec4f,
+    @location(4) color : vec4f,
+  }
+
+  @vertex
+  fn vert_main(
+    @location(0) a_particlePos : vec2f,
+    @location(1) a_particleVel : vec2f,
+    @location(2) a_pos : vec2f
+  ) -> VertexOutput {
+    let angle = -atan2(a_particleVel.x, a_particleVel.y);
+    let pos = vec2(
+      (a_pos.x * cos(angle)) - (a_pos.y * sin(angle)),
+      (a_pos.x * sin(angle)) + (a_pos.y * cos(angle))
+    );
+
+    var output : VertexOutput;
+    output.position = vec4(pos + a_particlePos, 0.0, 1.0);
+    output.color = vec4(
+      1.0 - sin(angle + 1.0) - a_particleVel.y,
+      pos.x * 100.0 - a_particleVel.y + 0.1,
+      a_particleVel.x + cos(angle + 0.5),
+      1.0);
+    return output;
+  }
+
+  @fragment
+  fn frag_main(@location(4) color : vec4f) -> @location(0) vec4f {
+    return color;
+  }
+  ```
+  </details>
+  </TabItem>
+</Tabs>
+
+If we wanted to make use of TypeGPU's type-safe features with these shaders, we could create the appropriate definitions by hand.
+This, however, can be a tedious and difficult to maintain process.
```

Comment:
maybe we could signal somehow more clearly that this is only tedious when migrating existing large projects and not that using our library is `tedious and difficult to maintain`